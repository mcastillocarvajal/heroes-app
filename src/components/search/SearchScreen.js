import React, { useMemo } from 'react'
import queryString from 'query-string';
import { useLocation } from 'react-router';
import { heroes } from '../../data/heroes'
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({ history }) => {

    const location = useLocation();
    const { q = '' } = queryString.parse( location.search );

    const initialForm = { searchText: q }

    const [ formValues, handleInputChange ] = useForm( initialForm )

    const { searchText } = formValues;

    const FilteredHeroes = useMemo(() => getHeroesByName( q ), [ q ]);

    const handleSearch = (e) => {
        e.preventDefault();

        history.push( `?q=${searchText}` );
    }

    return (
        <div>
            <h1>Search Screen</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr/>

                    <form onSubmit={ handleSearch }>
                        <input
                            type="text"
                            placeholder="Find your hero"
                            className="form-control"
                            autoComplete="off"
                            name="searchText"
                            value={ searchText }
                            onChange={ handleInputChange }
                        />
                        <button 
                            type="submit"
                            className="btn btn-outline-primary mt-2 w-100"
                        >
                            Search
                        </button>
                    </form>
                </div>

                <div className="col-7">
                    <h4>Results</h4>
                    <hr/>

                    {
                        ( q === '') 
                            && 
                            <div className="alert alert-warning">Search for a hero</div>
                    }

                    {
                        ( q !== '' && FilteredHeroes.length === 0 ) 
                            && 
                            <div className="alert alert-danger">There's no { q } hero</div>
                    }

                    {
                        FilteredHeroes.map( hero => (
                            <HeroCard 
                                key={ hero.id }
                                { ...hero }
                            />
                        ))
                    }
                </div>
            </div>
        </div>
    )
}