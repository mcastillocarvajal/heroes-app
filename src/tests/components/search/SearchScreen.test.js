import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('test for <SearchScreen />', () => {

    
    test('should show comp correctly with default values ', () => {
        
        const wrapper = mount( 
            <MemoryRouter initialEntries={[ '/search' ]}>
                <Route path='/search' component={ SearchScreen } />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-warning').text().trim() ).toBe('Search for a hero');
    });

    test('should show Batman and input with queryString value', () => {
        
        const wrapper = mount( 
            <MemoryRouter initialEntries={[ '/search?q=batman' ]}>
                <Route path='/search' component={ SearchScreen } />
            </MemoryRouter>
        );
        expect( wrapper.find('input').prop('value') ).toBe('batman');
        expect( wrapper ).toMatchSnapshot();
    })

    test('should show an error if there is no hero found', () => {
        
        const wrapper = mount( 
            <MemoryRouter initialEntries={[ '/search?q=batman123' ]}>
                <Route path='/search' component={ SearchScreen } />
            </MemoryRouter>
        );

        expect( wrapper.find('.alert-danger').exists() ).toBe(true);
        expect( wrapper ).toMatchSnapshot();
    });
    
    test('should call the history.push', () => {
        
        const history = {
            push: jest.fn()
        };

        const wrapper = mount( 
            <MemoryRouter initialEntries={[ '/search?q=batman' ]}>
                <Route 
                    path='/search' 
                    component={ () => <SearchScreen history={ history } /> }
                />
            </MemoryRouter>
        );
        
        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        expect( history.push ).toHaveBeenCalledWith('?q=batman')
    });
})
