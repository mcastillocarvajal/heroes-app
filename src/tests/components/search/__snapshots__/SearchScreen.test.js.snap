// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test for <SearchScreen /> should show Batman and input with queryString value 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary mt-2 w-100"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card mb-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row g-0"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p>
                <small
                  className="text-mutes"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                More...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`test for <SearchScreen /> should show an error if there is no hero found 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman123"
        />
        <button
          className="btn btn-outline-primary mt-2 w-100"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        There's no 
        batman123
         hero
      </div>
    </div>
  </div>
</div>
`;

exports[`test for <SearchScreen /> should show comp correctly with default values  1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-2 w-100"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-warning"
      >
        Search for a hero
      </div>
    </div>
  </div>
</div>
`;
